File: backend/app/schemas/review.py
"""Review and training dataset Pydantic schemas."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


# =============================================================================
# Annotation Statistics
# =============================================================================


class AnnotationClassStats(BaseModel):
    """Statistics for a single annotation class."""

    class_name: str
    class_color: str
    total_count: int
    frame_count: int  # Number of frames containing this class
    avg_confidence: float
    annotation_ids: list[str]  # For individual filtering


class AnnotationStatsResponse(BaseModel):
    """Aggregated annotation statistics for filtering UI."""

    job_id: str
    total_annotations: int
    total_frames: int
    classes: list[AnnotationClassStats]


# =============================================================================
# Frame Diversity Analysis
# =============================================================================


class DiversityAnalysisRequest(BaseModel):
    """Request to analyze frame diversity."""

    similarity_threshold: float = Field(
        default=0.85,
        ge=0.0,
        le=1.0,
        description="Frames with similarity > threshold are considered duplicates",
    )
    motion_threshold: float = Field(
        default=0.02,
        ge=0.0,
        le=1.0,
        description="Frames with motion < threshold are considered low-motion",
    )
    sample_camera: Literal["left", "right"] = "left"


class FrameCluster(BaseModel):
    """A cluster of similar frames."""

    representative_index: int  # The selected frame (first in cluster)
    member_indices: list[int]  # All frames in cluster
    avg_similarity: float


class DiversityAnalysisResponse(BaseModel):
    """Results of frame diversity analysis."""

    job_id: str
    status: Literal["pending", "analyzing", "complete", "failed"]
    error_message: str | None = None

    # Results (when complete)
    selected_frame_indices: list[int] = Field(default_factory=list)
    excluded_frame_indices: list[int] = Field(default_factory=list)
    clusters: list[FrameCluster] = Field(default_factory=list)

    # Statistics
    original_frame_count: int = 0
    selected_frame_count: int = 0
    reduction_percent: float = 0.0
    duplicate_pairs_found: int = 0
    low_motion_frames: int = 0

    # Per-frame data for visualization (only when requested)
    perceptual_hashes: dict[int, str] = Field(default_factory=dict)
    motion_scores: dict[int, float] = Field(default_factory=dict)


# =============================================================================
# Filter Configuration
# =============================================================================


class FilterConfiguration(BaseModel):
    """User's filter selections for training dataset export."""

    excluded_classes: list[str] = Field(default_factory=list)
    excluded_annotation_ids: list[str] = Field(default_factory=list)
    excluded_frame_indices: list[int] = Field(default_factory=list)
    diversity_applied: bool = False
    similarity_threshold: float | None = None
    motion_threshold: float | None = None


# =============================================================================
# Training Dataset
# =============================================================================


class TrainingDatasetRequest(BaseModel):
    """Request to create a training dataset."""

    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    format: Literal["kitti", "coco", "both"] = "both"

    # Filtering configuration
    filter_config: FilterConfiguration = Field(default_factory=FilterConfiguration)

    # Split configuration
    train_ratio: float = Field(default=0.7, ge=0.0, le=1.0)
    val_ratio: float = Field(default=0.2, ge=0.0, le=1.0)
    test_ratio: float = Field(default=0.1, ge=0.0, le=1.0)
    shuffle_seed: int | None = 42

    # Export options
    include_masks: bool = True
    include_depth: bool = True
    include_3d_boxes: bool = False


class TrainingDatasetResponse(BaseModel):
    """Response from training dataset creation."""

    id: UUID
    job_id: UUID
    name: str
    status: Literal["pending", "processing", "complete", "failed"]
    progress: float = 0.0

    # Statistics
    total_frames: int = 0
    total_annotations: int = 0
    train_count: int = 0
    val_count: int = 0
    test_count: int = 0

    created_at: datetime


class TrainingDatasetDetail(TrainingDatasetResponse):
    """Detailed training dataset information with lineage."""

    description: str | None = None
    format: str
    filter_config: FilterConfiguration

    # Lineage
    source_job_id: UUID
    source_job_name: str | None = None
    source_dataset_id: UUID | None = None
    source_dataset_name: str | None = None

    # Output
    output_directory: str | None = None
    kitti_path: str | None = None
    coco_path: str | None = None
    file_size_bytes: int | None = None

    # Timing
    completed_at: datetime | None = None
    error_message: str | None = None


class TrainingDatasetListResponse(BaseModel):
    """Response schema for training dataset list."""

    datasets: list[TrainingDatasetResponse]
    total: int


# =============================================================================
# Frame Batch for Playback
# =============================================================================


class FrameBatchRequest(BaseModel):
    """Request for batch of frames for video playback."""

    start_index: int = Field(ge=0)
    count: int = Field(default=24, ge=1, le=100)


class FrameThumbnail(BaseModel):
    """Minimal frame data for playback."""

    frame_id: str
    sequence_index: int
    svo2_frame_index: int
    thumbnail_url: str
    annotation_count: int


class FrameBatchResponse(BaseModel):
    """Response with batch of frames for video playback."""

    job_id: str
    frames: list[FrameThumbnail]
    total_frames: int
    start_index: int
    has_more: bool

File: frontend/src/types/review.ts
/**
 * Review mode and training dataset types.
 */

// =============================================================================
// Annotation Statistics
// =============================================================================

export interface AnnotationClassStats {
  class_name: string;
  class_color: string;
  total_count: number;
  frame_count: number;
  avg_confidence: number;
  annotation_ids: string[];
}

export interface AnnotationStatsResponse {
  job_id: string;
  total_annotations: number;
  total_frames: number;
  classes: AnnotationClassStats[];
}

// =============================================================================
// Frame Diversity Analysis
// =============================================================================

export interface DiversityAnalysisRequest {
  similarity_threshold: number;
  motion_threshold: number;
  sample_camera: 'left' | 'right';
}

export interface FrameCluster {
  representative_index: number;
  member_indices: number[];
  avg_similarity: number;
}

export interface DiversityAnalysisResponse {
  job_id: string;
  status: 'pending' | 'analyzing' | 'complete' | 'failed';
  error_message?: string;
  selected_frame_indices: number[];
  excluded_frame_indices: number[];
  clusters: FrameCluster[];
  original_frame_count: number;
  selected_frame_count: number;
  reduction_percent: number;
  duplicate_pairs_found: number;
  low_motion_frames: number;
  perceptual_hashes: Record<number, string>;
  motion_scores: Record<number, number>;
}

// =============================================================================
// Filter Configuration
// =============================================================================

export interface FilterConfiguration {
  excluded_classes: string[];
  excluded_annotation_ids: string[];
  excluded_frame_indices: number[];
  diversity_applied: boolean;
  similarity_threshold: number | null;
  motion_threshold: number | null;
}

// =============================================================================
// Training Dataset
// =============================================================================

export interface TrainingDatasetRequest {
  name: string;
  description?: string;
  format: 'kitti' | 'coco' | 'both';
  filter_config: FilterConfiguration;
  train_ratio: number;
  val_ratio: number;
  test_ratio: number;
  shuffle_seed?: number;
  include_masks: boolean;
  include_depth: boolean;
  include_3d_boxes: boolean;
}

export interface TrainingDatasetResponse {
  id: string;
  job_id: string;
  name: string;
  status: 'pending' | 'processing' | 'complete' | 'failed';
  progress: number;
  total_frames: number;
  total_annotations: number;
  train_count: number;
  val_count: number;
  test_count: number;
  created_at: string;
}

export interface TrainingDatasetDetail extends TrainingDatasetResponse {
  description: string | null;
  format: string;
  filter_config: FilterConfiguration;
  source_job_id: string;
  source_job_name: string | null;
  source_dataset_id: string | null;
  source_dataset_name: string | null;
  output_directory: string | null;
  kitti_path: string | null;
  coco_path: string | null;
  file_size_bytes: number | null;
  completed_at: string | null;
  error_message: string | null;
}

export interface TrainingDatasetListResponse {
  datasets: TrainingDatasetResponse[];
  total: number;
}

// =============================================================================
// Frame Batch for Playback
// =============================================================================

export interface FrameThumbnail {
  frame_id: string;
  sequence_index: number;
  svo2_frame_index: number;
  thumbnail_url: string;
  annotation_count: number;
}

export interface FrameBatchResponse {
  job_id: string;
  frames: FrameThumbnail[];
  total_frames: number;
  start_index: number;
  has_more: boolean;
}

// =============================================================================
// Playback State
// =============================================================================

export type PlaybackSpeed = 0.25 | 0.5 | 1 | 2 | 4;

export interface PlaybackState {
  isPlaying: boolean;
  speed: PlaybackSpeed;
  currentFrameIndex: number;
}
